Mitigate node out of disk status oscillation.

madhusudancs opened the pull request:
Fix for issue #12041.

madhusudancs commented:
Did not know Github allows only one assignee per PR.

@davidopp @lavalamp PTAL.

davidopp commented:
IIUC this PR will delay reporting
nodeOODCondition.Status = api.ConditionTrue
if you transitioned to reporting
nodeOODCondition.Status = api.ConditionFalse
less than 5 minutes ago. (And also the opposite transition.)

But I think this is not good -- I agree we want to delay reporting
nodeOODCondition.Status = api.ConditionFalse
if we transitioned to reporting
nodeOODCondition.Status = api.ConditionTrue
less than 5 minutes ago, but I do not think we ever want to 
delay reporting
nodeOODCondition.Status = api.ConditionTrue

In other words, I think we want to report "out of disk" as soon as it happens, so scheduler can immediately start avoiding that node, and only delay reporting "not out of disk." Does this make sense?

I think this is a simple change, just move the 
`if kl.clock.Since(nodeOODCondition.LastTransitionTime.Time) >= kl.outOfDiskTransitionFrequency {`
test into the branch that sets nodeOODCondition.Status = api.ConditionFalse
(and also update any relevant comments and flag description).

davidopp commented on docs/admin/kubelet.md:
```diff
@@ -113,6 +113,7 @@ kubelet
       --network-plugin-dir="/usr/libexec/kubernetes/kubelet-plugins/net/exec/": <Warning: Alpha feature> The full path of the directory in which to search for network plugins
       --node-status-update-frequency=10s: Specifies how often kubelet posts node status to master. Note: be cautious when changing the constant, it must work with nodeMonitorGracePeriod in nodecontroller. Default: 10s
       --oom-score-adj=-999: The oom-score-adj value for kubelet process. Values must be within the range [-1000, 1000]
+      --outofdisk-transition-frequency=5m0s: Duration for which the kubelet has to wait before transitioning from one out of disk node condition status to another. Default: 300s
```
elsewhere you wrote 300s

madhusudancs replied:
Made it 300s.

madhusudancs commented:
@davidopp Addressed the comment. PTAL.

madhusudancs commented:
@k8s-bot test this please.

davidopp commented on docs/admin/kubelet.md:
```diff
@@ -113,6 +113,7 @@ kubelet
       --network-plugin-dir="/usr/libexec/kubernetes/kubelet-plugins/net/exec/": <Warning: Alpha feature> The full path of the directory in which to search for network plugins
       --node-status-update-frequency=10s: Specifies how often kubelet posts node status to master. Note: be cautious when changing the constant, it must work with nodeMonitorGracePeriod in nodecontroller. Default: 10s
       --oom-score-adj=-999: The oom-score-adj value for kubelet process. Values must be within the range [-1000, 1000]
+      --outofdisk-transition-frequency=5m0s: Duration for which the kubelet has to wait before transitioning out of out-of-disk node condition status. Default: 300s
```
I would suggest doing 300s instead of 5m0s just for consistency.

madhusudancs replied:
I tried changing this before, but then I realized that this is an auto-generated file. It seems to parse 300s and automatically write it out as 5m0s. I have changed all the relevant occurrences to 5m0s for consistency.

davidopp commented on pkg/kubelet/kubelet.go:
```diff
@@ -320,7 +320,8 @@ func NewMainKubelet(
 		daemonEndpoints:                daemonEndpoints,
-		containerManager:               containerManager,
+		clock:                          util.RealClock{},
+		outOfDiskTransitionFrequency: outOfDiskTransitionFrequency,
```
this appears to be unaligned; you probably need to run gofmt

madhusudancs replied:
I think this is a gofmt bug. Since gofmt is run every time I save this file in the editor, this strange alignment keeps coming back. I even tried opening this file in an editor without gofmt integration, manually adjusting the alignment and then running gofmt on the command line again. This unaligned formatting just keeps happening.

davidopp commented on pkg/kubelet/kubelet.go:
```diff
@@ -637,6 +638,17 @@ type Kubelet struct {
+	// easy to test the code.
+	clock util.Clock
+
+	// outOfDiskTransitionFrequency specifies the duration for which the kubelet has to wait
```
This comment seems a bit overly verbose. I would say something like "outOfDiskTransitionFrequency specifies the amount of time the kubelet has to be actually not out of disk before it can transition the node condition status from out-of-disk to not-out-of-disk. This prevents a pod that causes out-of-disk condition from repeatedly getting rescheduled onto the node."

madhusudancs replied:
Done.

davidopp commented on pkg/kubelet/kubelet.go:
```diff
@@ -2731,7 +2741,7 @@ func (kl *Kubelet) setNodeStatus(node *api.Node) error {
 		}
 	} else {
-		if nodeOODCondition.Status != api.ConditionFalse {
+		if nodeOODCondition.Status != api.ConditionFalse && kl.clock.Since(nodeOODCondition.LastTransitionTime.Time) >= kl.outOfDiskTransitionFrequency {
```
maybe log a message if
nodeOODCondition.Status != api.ConditionFalse
but the clock condition is not true yet?

davidopp replied:
I think we should skip the clock check if nodeOODCondition.Status == api.ConditionUnknown. If we don't do that, I think on startup we will not start reporting nodeOODCondition.Status == api.ConditionFalse for five minutes, since the field is initialized to api.ConditionUnknown at startup. Same problem if the node is temporarily unreachable and NodeController marks it as nodeOODCondition.Status == api.ConditionUnknown. Do you agree?

madhusudancs replied:
> maybe log a message if nodeOODCondition.Status != api.ConditionFalse but the clock condition is not true yet?

Done.

> I think we should skip the clock check if nodeOODCondition.Status == api.ConditionUnknown. If we don't do that, I think on startup we will not start reporting nodeOODCondition.Status == api.ConditionFalse for five minutes, since the field is initialized to api.ConditionUnknown at startup. Same problem if the node is temporarily unreachable and NodeController marks it as nodeOODCondition.Status == api.ConditionUnknown. Do you agree?

Agreed. And fixed.

davidopp commented on pkg/kubelet/kubelet_test.go:
```diff
@@ -2926,6 +2993,166 @@ func TestUpdateExistingNodeStatus(t *testing.T) {
+
+		var oodCondition api.NodeCondition
+		for i, cond := range updatedNode.Status.Conditions {
+			// Expect LastProbeTime to be updated to Now, while LastTransitionTime to be the same.
```
This comment should be "Expect LastHeartbeatTime to be updated to Now"

madhusudancs replied:
Done.

davidopp commented on pkg/kubelet/kubelet_test.go:
```diff
@@ -2926,6 +2993,166 @@ func TestUpdateExistingNodeStatus(t *testing.T) {
+					t.Errorf("%d. Condition %v LastTransitionTime:\nexpected \n%#v\n, got \n%#v", tcIdx, cond.Type, clock.Now(), cond.LastTransitionTime.Time)
+				}
+			}
+			updatedNode.Status.Conditions[i].LastHeartbeatTime = unversioned.Time{}
```
I'm sure I'm missing someting, but I didn't understand the purpose of this line and the next line. Don't we want LastHeartbeatTime and LastTransitionTime to be set by the Kubelet, not by the test itself?

madhusudancs replied:
A few lines below, we do a deep equals between the objects. It is hard to compare time values.  So I followed this pattern that other tests use in this file.

Now that we can precisely control the time set by the tests, we can switch to deep equals based comparisons even for time values.

madhusudancs commented:
@davidopp PTAL.

vishh commented:
Given the fact that kubernetes already has low and high thresholds for disk reclamation, why do we think that kubelet will oscillate? 
The current defaults as percentage of disk capacity are 90% for high threshold and 60% for low threshold.
Kubelet attempts to reclaim disk space until it hits the low watermark.

davidopp commented:
Needs rebase.

@vishh The scenario we're concerned about is a small machine where we try to schedule a container with a large image that causes the machine to immediately go OOD; so the pod fails, and the machine immediately goes back to not-OOD once the image is GCd, so the scheduler re-tries that pod on that same machine again, etc. in an infinite loop. #12041 has more details.

davidopp commented:
LGTM

davidopp commented:
(but needs rebase)

madhusudancs commented:
Rebased. PTAL.

davidopp commented:
The changes to the files in docs/api-reference look suspicious. Were those intentional?

davidopp commented:
Also, please squash the commits.

madhusudancs commented:
> The changes to the files in docs/api-reference look suspicious. Were those intentional?

It wasn't intentional, it was a result of a rebase. Regenerated those files ago.

Squashed the commits. PTAL.

davidopp commented:
Wow, that's annoying that the auto-generators produce so much extra change to a PR.

Anyway, I've marked LGTM. Thanks!

wojtek-t commented on pkg/kubectl/testing/types.generated.go:
```diff
@@ -158,7 +158,13 @@ func (x *TestStruct) CodecEncodeSelf(e *codec1978.Encoder) {
 				if yyq2[2] {
 					yy10 := &x.ObjectMeta
-					yy10.CodecEncodeSelf(e)
+					yym11 := z.EncBinary()
```
Why do you change those auto-generated types? Your changes seem incorrect and are reason for test failures...

madhusudancs replied:
I did not change these. This is what hack/update-all.sh generated. And the pre-commits were passing.

madhusudancs replied:
So I ran hack/update-all.sh again, which runs hack/update-codecgen and I don't get any more diffs. I am not entirely sure what is going on.

madhusudancs committed:
Mitigate node out of disk status oscillation by delaying it.

Implement a flag that defines the frequency at which a node's out of
disk condition can change its status. Use this flag to suspend out of
disk status changes in the time period specified by the flag, after
the status is changed once.

Set the flag to 0 in e2e tests so that we can predictably test out of
disk node condition.

Also, use util.Clock interface for all time related functionality in
the kubelet. Calling time functions in unversioned package or time
package such as unversioned.Now() or time.Now() makes it really hard
to test such code. It also makes the tests flaky and sometimes
unnecessarily slow due to time.Sleep() calls used to simulate the
time elapsed. So use util.Clock interface instead which can be faked
in the tests.

madhusudancs commented:
@davidopp For an unrelated experiment, I had upgraded Go to 1.5.2 and that likely caused those go2idl testdata types to be generated. A similar behavior has also been reported in #18907. I downgraded Go to 1.4.3 and the generated code looks Ok now. I have also rebased the PR again. PTAL.

davidopp commented:
LGTM

I'm guessing the test failures are related to the problems we're having with the tests lately and not with this PR, but I didn't check.

madhusudancs commented:
@k8s-bot test this please.

madhusudancs commented:
@k8s-bot test this please.

madhusudancs commented:
@k8s-bot test this please.
